test/test_incorrect_dispatch_key_ops.py

```diff
@@ -0,0 +1,112 @@
+
+# The test written below is pretty fragile; it parses the aten/ folder to look for all
+# dispatch stub kernels, and maps them to native_functions.yaml ops through naming conventions.
+# I had to add a list of ignored ops here to avoid false positives.
+# For example, `at::native::cumprod` is a (valid) CompositeExplicitAutograd kernel, which is implemented
+# by calling directly into `at::native::_cumprod` - which has a CPU/CUDA kernel that calls `cumprod_stub()`,
+# which is registered to DispatchStub.
+# This is a kind of weird pattern, but it's valid, and hard to avoid catching in this test.
+#
+# If you create a future op that's implemented similarly and you're sure it's a false positive,
+# add it to this list.
+IGNORED_OPS = [
+    "cumprod",
+    "cumsum",
+    "logcumsumexp",
+]
+
+class TestIncorrectDispatchKeyOps(TestCase):
+
+    def find_dispatch_stub_kernels(self) -> List[str]:
+        curr_path = os.path.dirname(os.path.realpath(__file__))
+        root_dir = os.path.join(curr_path, '../aten/src/ATen/')
+        regex = re.compile("DEFINE_DISPATCH\((.*)_stub\)")
+        f_path = os.path.join(curr_path, '../aten/src/ATen/native/BinaryOps.cpp')
+
+        matches = []
+        for root, dirs, files in os.walk(root_dir):
+            for f_name in files:
+                if not f_name.endswith(".cpp"):
+                    continue
+                f_path = os.path.join(root, f_name)
+                with open(f_path, "r") as f:
+                    f_str = f.read()
+                    matches += regex.findall(f_str)
+
+        return matches
+
+    def test_find_invalid_CompositeExplicitAutograd_ops(self):
+
+        def op_name(f: NativeFunction) -> str:
+            op_name = f.func.name
+            op_name_str = str(op_name.name)
+            if op_name.overload_name:
+                op_name_str += f'_{op_name.overload_name}'
+            return op_name_str
+
+        def invalid(op: str) -> bool:
+            return op in dispatch_stub_kernels and op not in IGNORED_OPS
+
+        curr_path = os.path.dirname(os.path.realpath(__file__))
+        yaml_path = os.path.join(curr_path, '../aten/src/ATen/native/native_functions.yaml')
+
+        parsed_yaml = parse_native_yaml(yaml_path)
+        native_functions, backend_indices = parsed_yaml.native_functions, parsed_yaml.backend_indices
+        grouped_native_functions = get_grouped_native_functions(native_functions)
+
+        dispatch_stub_kernels = self.find_dispatch_stub_kernels()
+
+        invalid_op_entries: List[str] = []
+        for g in grouped_native_functions:
+            if isinstance(g, NativeFunction):
+                f = g
+                if not backend_indices[DispatchKey.CompositeExplicitAutograd].has_kernel(f):
+                    continue
+                op_name_str = op_name(f)
+                if invalid(op_name_str):
+                    invalid_op_entries.append(op_name_str)
+
+            elif isinstance(g, NativeFunctionsGroup):
```

Comment:
A little confused here, don't you just want to run the check on a flat list of NativeFunctions (i.e. just don't bother grouping them)

Reply:
I actually did that originally, but it causes the test to pick up way more false-positives.

For example, the native_functions.yaml entries for `at::abs` are:
```
- func: abs(Tensor self) -> Tensor
  dispatch:
    CompositeExplicitAutograd: abs
- func: abs_(Tensor(a!) self) -> Tensor(a!)
  dispatch:
    CompositeExplicitAutograd: abs_
- func: abs.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
  dispatch:
    CPU, CUDA: abs_out
```

It's valid for `abs` and `abs_` to be CompositeExplicitAutograd, and for `abs_out` to have backend-specific kernels that are implemented using DispatchStub. But the test would pick that up because it sees `DEFINE_DISPATCH(abs_stub`, along with a CompositeExplicitAutograd `abs` entry somewhere. This seemed like a common enough case that I just skip NativeFunctionGroup entries with a mix of composite and non-composite dispatch tables.

Reply:
Hey, maybe we could enforce a naming convention (through this test) that if an `{op}_out` kernel uses DispatchStub, it has to name the corresponding dispatch kernel `{op}_out_stub`. But that just seems even more intrusive.